File: 287_FindTheDuplicateNumber.py
##30.25%
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        i, size = 0, len(nums)
        while i < size:
            while nums[i] != i + 1:
                index = nums[i]
                if nums[i] == nums[index - 1]:
                    return nums[i]
                nums[i], nums[index - 1] = nums[index - 1], nums[i]
            i += 1
        return nums[-1]
		
def findArrayDuplicate(array):
    assert len(array) > 0

    # The "tortoise and hare" step.  We start at the end of the array and try
    # to find an intersection point in the cycle.
    slow = len(array) - 1
    fast = len(array) - 1

    # Keep advancing 'slow' by one step and 'fast' by two steps until they
    # meet inside the loop.
    while True:
        slow = array[slow]
        fast = array[array[fast]]

        if slow == fast:
            break

    # Start up another pointer from the end of the array and march it forward
    # until it hits the pointer inside the array.
    finder = len(array) - 1
    while True:
        slow   = array[slow]
        finder = array[finder]

        # If the two hit, the intersection index is the duplicate element.
        if slow == finder:
            return slow

File: 287_FindTheDuplicateNumber.cpp
//58.17%
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int size(nums.size());
		for(int i = 0; i < size; ++i){
			while(nums[i] != i + 1){
				if(nums[nums[i] - 1] == nums[i])
					return nums[i];
				swap(nums[i], nums[nums[i] - 1]);
			}
		}
		return nums[size - 1];
    }
};

//Wonderful solution which didn't change input variable parameters
int findDuplicate3(vector<int>& nums)
{
	if (nums.size() > 1)
	{
		int slow = nums[0];
		int fast = nums[nums[0]];
		while (slow != fast)
		{
			slow = nums[slow];
			fast = nums[nums[fast]];
		}

		fast = 0;
		while (fast != slow)
		{
			fast = nums[fast];
			slow = nums[slow];
		}
		return slow;
	}
	return -1;
}

class Solution {
public:
int findDuplicate(vector<int>& nums) {
    int slow = 0;
	int fast = 0;
	int finder = 0;

	while (true)
	{
		slow = nums[slow];
		fast = nums[nums[fast]];

		if (slow == fast)
			break;
	}
	while (true)
	{
		slow = nums[slow];
		finder = nums[finder];
		if (slow == finder)
			return slow;
	}
}